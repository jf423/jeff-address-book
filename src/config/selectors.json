{
    "INDEX_PAGE": {
        "INDEX_CONTAINER": {
            "ID": "INDEX_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='INDEX_CONTAINER']",
            "EX": "[data-selector-id='INDEX_CONTAINER']",
            "PARAM": ""
        },
        "SEARCH_CONTAINER": {
            "ID": "SEARCH_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_CONTAINER']",
            "EX": "[data-selector-id='SEARCH_CONTAINER']",
            "PARAM": ""
        },
        "SEARCH_INPUT": {
            "ID": "SEARCH_INPUT",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_INPUT']",
            "EX": "[data-selector-id='SEARCH_INPUT']",
            "PARAM": ""
        },
        "SEARCH_CLEAR": {
            "ID": "SEARCH_CLEAR",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_CLEAR']",
            "EX": "[data-selector-id='SEARCH_CLEAR']",
            "PARAM": ""
        },
        "SEARCH_BUTTON": {
            "ID": "SEARCH_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_BUTTON']",
            "EX": "[data-selector-id='SEARCH_BUTTON']",
            "PARAM": ""
        },
        "USER_CONTAINER": {
            "ID": "USER_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CONTAINER']",
            "EX": "[data-selector-id='USER_CONTAINER']",
            "PARAM": ""
        },
        "USER_CONTENT": {
            "ID": "USER_CONTENT",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CONTENT']",
            "EX": "[data-selector-id='USER_CONTENT']",
            "PARAM": ""
        },
        "USER_CATALOG_END": {
            "ID": "USER_CATALOG_END",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CATALOG_END']",
            "EX": "[data-selector-id='USER_CATALOG_END']",
            "PARAM": ""
        },
        "USER_CELL_CONTAINER": {
            "ID": "USER_CELL_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1']",
            "PARAM": ""
        },
        "USER_CELL_CONTENT": {
            "ID": "USER_CELL_CONTENT",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_CONTENT']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_CONTENT']",
            "PARAM": ""
        },
        "USER_CELL_AVATAR": {
            "ID": "USER_CELL_AVATAR",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_AVATAR']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_AVATAR']",
            "PARAM": ""
        },
        "USER_CELL_NAME": {
            "ID": "USER_CELL_NAME",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_NAME']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_NAME']",
            "PARAM": ""
        },
        "USER_CELL_NATIONALITY": {
            "ID": "USER_CELL_NATIONALITY",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_NATIONALITY']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_NATIONALITY']",
            "PARAM": ""
        },
        "USER_CELL_USERNAME": {
            "ID": "USER_CELL_USERNAME",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_USERNAME']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_USERNAME']",
            "PARAM": ""
        },
        "USER_CELL_EMAIL": {
            "ID": "USER_CELL_EMAIL",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_EMAIL']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_EMAIL']",
            "PARAM": ""
        },
        "USER_MODAL_CONTAINER": {
            "ID": "USER_MODAL_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CONTAINER']",
            "EX": "[data-selector-id='USER_MODAL_CONTAINER']",
            "PARAM": ""
        },
        "USER_MODAL_CLOSE_BUTTON": {
            "ID": "USER_MODAL_CLOSE_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CLOSE_BUTTON']",
            "EX": "[data-selector-id='USER_MODAL_CLOSE_BUTTON']",
            "PARAM": ""
        },
        "USER_MODAL_AVATAR": {
            "ID": "USER_MODAL_AVATAR",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_AVATAR']",
            "EX": "[data-selector-id='USER_MODAL_AVATAR']",
            "PARAM": ""
        },
        "USER_MODAL_NAME": {
            "ID": "USER_MODAL_NAME",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_NAME']",
            "EX": "[data-selector-id='USER_MODAL_NAME']",
            "PARAM": ""
        },
        "USER_MODAL_STREET": {
            "ID": "USER_MODAL_STREET",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_STREET']",
            "EX": "[data-selector-id='USER_MODAL_STREET']",
            "PARAM": ""
        },
        "USER_MODAL_CITY": {
            "ID": "USER_MODAL_CITY",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CITY']",
            "EX": "[data-selector-id='USER_MODAL_CITY']",
            "PARAM": ""
        },
        "USER_MODAL_STATE": {
            "ID": "USER_MODAL_STATE",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_STATE']",
            "EX": "[data-selector-id='USER_MODAL_STATE']",
            "PARAM": ""
        },
        "USER_MODAL_POSTCODE": {
            "ID": "USER_MODAL_POSTCODE",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_POSTCODE']",
            "EX": "[data-selector-id='USER_MODAL_POSTCODE']",
            "PARAM": ""
        },
        "USER_MODAL_PHONE": {
            "ID": "USER_MODAL_PHONE",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_PHONE']",
            "EX": "[data-selector-id='USER_MODAL_PHONE']",
            "PARAM": ""
        },
        "USER_MODAL_CELL": {
            "ID": "USER_MODAL_CELL",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CELL']",
            "EX": "[data-selector-id='USER_MODAL_CELL']",
            "PARAM": ""
        }
    },
    "SETTINGS_PAGE": {
        "SETTINGS_CONTAINER": {
            "ID": "SETTINGS_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='SETTINGS_CONTAINER']",
            "EX": "[data-selector-id='SETTINGS_CONTAINER']",
            "PARAM": ""
        },
        "NATIONALITIES_CONTAINER": {
            "ID": "NATIONALITIES_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CONTAINER']",
            "EX": "[data-selector-id='NATIONALITIES_CONTAINER']",
            "PARAM": ""
        },
        "NATIONALITIES_BACK_BUTTON": {
            "ID": "NATIONALITIES_BACK_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_BACK_BUTTON']",
            "EX": "[data-selector-id='NATIONALITIES_BACK_BUTTON']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_CH": {
            "ID": "NATIONALITIES_CHECKBOX_CH",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_CH']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_CH']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_ES": {
            "ID": "NATIONALITIES_CHECKBOX_ES",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_ES']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_ES']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_FR": {
            "ID": "NATIONALITIES_CHECKBOX_FR",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_FR']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_FR']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_GB": {
            "ID": "NATIONALITIES_CHECKBOX_GB",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_GB']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_GB']",
            "PARAM": ""
        },
        "NATIONALITIES_SAVE_BUTTON": {
            "ID": "NATIONALITIES_SAVE_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_SAVE_BUTTON']",
            "EX": "[data-selector-id='NATIONALITIES_SAVE_BUTTON']",
            "PARAM": ""
        }
    }
}