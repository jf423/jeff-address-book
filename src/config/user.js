export const USERS_CONFIG = {
    NUM_COLUMNS: 4,
    BATCH_NUM: 50,
    MAXIMUM_COUNT: 150,
    ITEM_WIDTH: 250,
    ITEM_HEIGHT: 300
};

export const END_POINT = 'https://randomuser.me/api/';

export const DEFAULT_NATIONALITIES = ['ch', 'es', 'fr', 'gb'];
