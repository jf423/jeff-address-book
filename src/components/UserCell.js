import PropTypes from 'prop-types';
import styled from 'styled-components';

import { flexRowCenterCenter, flexColCenterCenter } from '../style/flex.css.js';
import { eclipse, shadow } from '../style/common.css.js';

import { USERS_CONFIG } from '../config/user.js';
import { INDEX_PAGE } from '../config/selectors.json';

const { NUM_COLUMNS } = USERS_CONFIG;
const {
    USER_CELL_CONTAINER,
    USER_CELL_CONTENT,
    USER_CELL_AVATAR,
    USER_CELL_NAME,
    USER_CELL_NATIONALITY,
    USER_CELL_USERNAME,
    USER_CELL_EMAIL
} = INDEX_PAGE;

const UserCell = ({
    data: { items, setSelectedIndex },
    columnIndex,
    rowIndex,
    style
}) => {
    const itemIndex = rowIndex * NUM_COLUMNS + columnIndex;
    const itemData = items[itemIndex];

    const onClick = () => {
        setSelectedIndex(itemIndex);
    };

    if (!itemData) {
        return null;
    }
    return (
        <Card data-selector-id={`${USER_CELL_CONTAINER.ID}_${itemIndex}`} style={style}>
            <CardContent data-selector-id={USER_CELL_CONTENT.ID} onClick={onClick}>
                <Avatar data-selector-id={USER_CELL_AVATAR.ID} src={itemData?.picture?.thumbnail} />
                <Name data-selector-id={USER_CELL_NAME.ID}>
                    { itemData.name ? `${itemData?.name?.first} ${itemData?.name?.last}` : 'Loading'}
                </Name>
                <Info>
                    <Text data-selector-id={USER_CELL_NATIONALITY.ID}>{itemData.nat && `(${itemData?.nat})`}</Text>
                    <Text data-selector-id={USER_CELL_USERNAME.ID}>{itemData?.login?.username}</Text>
                    <Text data-selector-id={USER_CELL_EMAIL.ID}>{itemData?.email}</Text>
                </Info>
            </CardContent>
        </Card>
    );
};

export default UserCell;

UserCell.propTypes = {
    data: PropTypes.shape({
        items: PropTypes.array.isRequired,
        setSelectedIndex: PropTypes.func.isRequired
    }).isRequired,
    columnIndex: PropTypes.number.isRequired,
    rowIndex: PropTypes.number.isRequired,
    style: PropTypes.object.isRequired
};

const Card = styled.div`
    ${flexRowCenterCenter};
`;

const CardContent = styled.div`
    position: relative;
    width: -webkit-fill-available;
    height: -webkit-fill-available;
    margin: 10px;
    padding: 10px;
    cursor: pointer;
    ${shadow};
    ${flexColCenterCenter};

    &:before {
        content: '';
        width: 100%;
        height: 75px;
        background-color: #b2b2b2;
        position: absolute;
        top: 0;
        left: 0;
        z-index: -1;
    }
`;

const Avatar = styled.img`
    border: 5px solid #fff;
    background: #ddd;
    border-radius: 100%;
    width: 100px;
    height: 100px;
`;

const Name = styled.h2`
    text-align: center;
    max-width: 100%;
    ${eclipse};
`;

const Info = styled.div`
    width: 100%;
    min-height: 80px;
    ${flexColCenterCenter};
`;

const Text = styled.span`
    max-width: 100%;
    padding: 5px;
    color: #6c757d;
    ${eclipse};
`;
