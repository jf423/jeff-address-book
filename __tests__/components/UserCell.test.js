import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { toBeInTheDocument, toHaveTextContent, toHaveValue } from '@testing-library/jest-dom/matchers';

import { USERS_CONFIG } from '../../src/config/user.js';
import { INDEX_PAGE } from '../../src/config/selectors.json';

import UserCell from '../../src/components/UserCell.js';

import DATA from '../mockUsersData.json';

expect.extend({ toBeInTheDocument, toHaveTextContent, toHaveValue });

const mockData = {
    data: {
        items: DATA.results,
        setSelectedIndex: jest.fn()
    },
    columnIndex: 0,
    rowIndex: 0,
    style: {}
};
const { NUM_COLUMNS } = USERS_CONFIG;
const {
    USER_CELL_CONTAINER,
    USER_CELL_CONTENT,
    USER_CELL_AVATAR,
    USER_CELL_NAME,
    USER_CELL_NATIONALITY,
    USER_CELL_USERNAME,
    USER_CELL_EMAIL
} = INDEX_PAGE;

describe('UserCell component test', () => {
    let userCell;
    const renderUserCell = props => (
        <UserCell {...props} />
    );

    beforeEach(() => {
        userCell = render(renderUserCell(mockData));
    });

    afterEach(() => {
        userCell.unmount();
    });

    test('should display as expected', () => {
        const itemIndex = mockData.rowIndex * NUM_COLUMNS + mockData.columnIndex;
        const item = DATA.results[itemIndex];
        const userCellContainer = userCell.container.querySelector(`[data-selector-id="${USER_CELL_CONTAINER.ID}_${itemIndex}"]`);
        const userCellContent = userCellContainer.querySelector(`[data-selector-id="${USER_CELL_CONTENT.ID}"]`);
        const userCellAvatar = userCellContainer.querySelector(`[data-selector-id="${USER_CELL_AVATAR.ID}"]`);
        const userCellName = userCellContainer.querySelector(`[data-selector-id="${USER_CELL_NAME.ID}"]`);
        const userCellNationality = userCellContainer.querySelector(`[data-selector-id="${USER_CELL_NATIONALITY.ID}"]`);
        const userCellUsername = userCellContainer.querySelector(`[data-selector-id="${USER_CELL_USERNAME.ID}"]`);
        const userCellEmail = userCellContainer.querySelector(`[data-selector-id="${USER_CELL_EMAIL.ID}"]`);

        expect(userCellContainer).toBeInTheDocument();
        expect(userCellContent).toBeInTheDocument();
        expect(userCellAvatar.src).toBe(item?.picture?.thumbnail);
        expect(userCellName).toHaveTextContent(`${item?.name?.first} ${item?.name?.last}`);
        expect(userCellNationality).toHaveTextContent(`(${item?.nat})`);
        expect(userCellUsername).toHaveTextContent(item?.login?.username);
        expect(userCellEmail).toHaveTextContent(item?.email);
        expect(userCell).toMatchSnapshot();
    });

    test('should be trigger correctly when the user clicks the cell', () => {
        const itemIndex = mockData.rowIndex * NUM_COLUMNS + mockData.columnIndex;
        const userCellContainer = userCell.container.querySelector(`[data-selector-id="${USER_CELL_CONTAINER.ID}_${itemIndex}"]`);
        const userCellContent = userCellContainer.querySelector(`[data-selector-id="${USER_CELL_CONTENT.ID}"]`);

        fireEvent.click(userCellContent);
        expect(mockData.data.setSelectedIndex).toHaveBeenCalledWith(itemIndex);
    });
});
