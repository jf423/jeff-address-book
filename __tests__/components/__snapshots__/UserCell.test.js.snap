// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCell component test should display as expected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="UserCell__Card-sc-1lznl0l-0 jPkKlV"
        data-selector-id="USER_CELL_CONTAINER_0"
      >
        <div
          class="UserCell__CardContent-sc-1lznl0l-1 jgliMZ"
          data-selector-id="USER_CELL_CONTENT"
        >
          <img
            class="UserCell__Avatar-sc-1lznl0l-2 trjPI"
            data-selector-id="USER_CELL_AVATAR"
            src="https://randomuser.me/api/portraits/thumb/men/29.jpg"
          />
          <h2
            class="UserCell__Name-sc-1lznl0l-3 EBjpy"
            data-selector-id="USER_CELL_NAME"
          >
            Mario Sanchez
          </h2>
          <div
            class="UserCell__Info-sc-1lznl0l-4 dOQZzn"
          >
            <span
              class="UserCell__Text-sc-1lznl0l-5 fxqfUb"
              data-selector-id="USER_CELL_NATIONALITY"
            >
              (ES)
            </span>
            <span
              class="UserCell__Text-sc-1lznl0l-5 fxqfUb"
              data-selector-id="USER_CELL_USERNAME"
            >
              ticklishleopard646
            </span>
            <span
              class="UserCell__Text-sc-1lznl0l-5 fxqfUb"
              data-selector-id="USER_CELL_EMAIL"
            >
              mario.sanchez@example.com
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="UserCell__Card-sc-1lznl0l-0 jPkKlV"
      data-selector-id="USER_CELL_CONTAINER_0"
    >
      <div
        class="UserCell__CardContent-sc-1lznl0l-1 jgliMZ"
        data-selector-id="USER_CELL_CONTENT"
      >
        <img
          class="UserCell__Avatar-sc-1lznl0l-2 trjPI"
          data-selector-id="USER_CELL_AVATAR"
          src="https://randomuser.me/api/portraits/thumb/men/29.jpg"
        />
        <h2
          class="UserCell__Name-sc-1lznl0l-3 EBjpy"
          data-selector-id="USER_CELL_NAME"
        >
          Mario Sanchez
        </h2>
        <div
          class="UserCell__Info-sc-1lznl0l-4 dOQZzn"
        >
          <span
            class="UserCell__Text-sc-1lznl0l-5 fxqfUb"
            data-selector-id="USER_CELL_NATIONALITY"
          >
            (ES)
          </span>
          <span
            class="UserCell__Text-sc-1lznl0l-5 fxqfUb"
            data-selector-id="USER_CELL_USERNAME"
          >
            ticklishleopard646
          </span>
          <span
            class="UserCell__Text-sc-1lznl0l-5 fxqfUb"
            data-selector-id="USER_CELL_EMAIL"
          >
            mario.sanchez@example.com
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
